src/settings.cpp

```diff
@@ -33,48 +33,100 @@ with this program; if not, write to the Free Software Foundation, Inc.,
 #include <cctype>
 #include <algorithm>
 
-Settings *g_settings = nullptr; // Populated in main()
+Settings *g_settings = nullptr;
+static SettingsHierarchy g_hierarchy;
 std::string g_settings_path;
 
-Settings *Settings::s_layers[SL_TOTAL_COUNT] = {0}; // Zeroed by compiler
 std::unordered_map<std::string, const FlagDesc *> Settings::s_flags;
 
+/* Settings hierarchy implementation */
 
-Settings *Settings::createLayer(SettingsLayer sl, const std::string &end_tag)
+SettingsHierarchy::SettingsHierarchy(Settings *fallback)
+{
+	layers.resize(1);
+	layers[0] = fallback;
+}
+
+
+Settings *SettingsHierarchy::getLayer(int layer) const
 {
-	if ((int)sl < 0 || sl >= SL_TOTAL_COUNT)
+	if (layer < 0 || layer >= layers.size())
 		throw BaseException("Invalid settings layer");
+	return layers[layer];
+}
+
 
-	Settings *&pos = s_layers[(size_t)sl];
+Settings *SettingsHierarchy::getParent(int layer) const
+{
+	assert(layer >= 0 && layer < layers.size());
+	// iterate towards the origin (0) to find the next fallback layer
+	for (int i = layer - 1; i >= 0; --i) {
+		if (layers[i])
+			return layers[i];
+	}
+
+	return nullptr;
+}
+
+
+void SettingsHierarchy::layerCreated(int layer, Settings *obj)
+{
+	if (layer < 0)
+		throw BaseException("Invalid settings layer");
+	if (layers.size() < layer+1)
+		layers.resize(layer+1);
+
+	Settings *&pos = layers[layer];
 	if (pos)
-		throw BaseException("Setting layer " + std::to_string(sl) + " already exists");
+		throw BaseException("Setting layer " + itos(layer) + " already exists");
+
+	pos = obj;
+	// This feels bad
+	if (this == &g_hierarchy && layer == (int)SL_GLOBAL)
+		g_settings = obj;
+}
+
+
+void SettingsHierarchy::layerRemoved(int layer)
```

Comment:
u32

Reply:
`onLayerRemove`. The instance is yet not deleted.